src/lib/domain.js

```diff
@@ -0,0 +1,122 @@
+/**
+ * Extracts `hostname`, `pathname`, `isWildcard` from a given `domain`
+ *
+ * @param {string} domain
+ * @return {{ hostname: string, pathname: string, isWildcard: boolean }}
+ */
+export function parseDomain (domain) {
+  const domainWithHttp = domain.match(/^https?:\/\//) != null ? domain : 'https://' + domain;
+  // With firefox, 'https://*.toto.com' is considered invalid so we strip it off for the test
+  // because we know this part is valid and we want the rest to be sanitized by the URL parser
+  const isWildcard = domain.startsWith('*.');
+  try {
+    const { hostname, pathname } = new URL(domainWithHttp.replace('*.', ''));
+
+    return {
+      hostname,
+      pathname,
+      isWildcard,
+    };
+  }
+  catch {
+    if (domain.length === 0) {
+      throw new DomainError('empty');
+    }
+
+    if (!isWildcard && domain.includes('*')) {
+      throw new DomainError('invalid-wildcard');
+    }
+
+    throw new DomainError('invalid-format');
+  }
+}
+
+export class DomainError extends Error {
+  /** @param {'empty'|'invalid-wildcard'|'invalid-format'} message */
+  constructor (message) {
+    super();
+
+    this.message = message;
+  }
+}
+
+/**
+ * If isWildcard is `true`, returns `*.${hostname}`
+ * else returns the `hostname` untouched
+ *
+ * @param {string} hostname
+ * @param {boolean} isWildcard
+ * @returns {string}
+ */
+export function getHostWithWildcard (hostname, isWildcard) {
+  return [
+    isWildcard ? '*.' : '',
+    hostname,
+  ].join('');
+}
+
+/**
+ * @param {string} hostname
+ * @param {string} pathPrefix
+ * @param {boolean} isWildcard
+ * @param {boolean} isHttpOnly
+ * @returns {string}
+ */
+export function getDomainUrl (hostname, pathPrefix, isWildcard, isHttpOnly) {
+  return [
+    isHttpOnly ? 'http://' : 'https://',
+    isWildcard ? 'www.' : '',
+    hostname,
+    pathPrefix,
+  ].join('');
+}
+
+/**
+ * Checks if the domain is a `cleverapps.io` domain and if it contains a subdomain.
+ * For instance `subdomain.main.cleverapps.io` is HTTP only.
+ *
+ * @param {string} domain
+ * @return {boolean} whether the domain is a `cleverapps.io` is HTTP only or not
+ */
+export function isTestDomainWithSubdomain (domain) {
+  return domain.endsWith('cleverapps.io') && domain.split('.').length > 3;
```

Comment:
:+1: for using `isTestDomain` but this means keeping the dot split instead of the regex (I think it's OK). Also, we should always have a `$` at the end for such regex.

Reply:
I'll stick with `isTestDomain` + split.
We'll have to revisit this code if / when we add new test domains in the future anyway